go/pkg/client/tree.go

```diff
@@ -87,115 +87,127 @@ func shouldIgnore(inp string, t command.InputType, excl []*command.InputExclusio
-			},
+
+	for len(filesToProcess) != 0 {
+		path := filesToProcess[0]
+		filesToProcess = filesToProcess[1:]
+
+		if path == "" {
+			return errors.New("empty Input, use \".\" for entire exec root")
 		}
-		return nil
-	} else if t == command.SymlinkInputType {
-		relTarget, err := getTargetRelPath(execRoot, symlinkNormDir, meta.Symlink.Target)
+
+		absPath := filepath.Join(execRoot, path)
+
+		meta := cache.Get(absPath)
+
+		normPath, err := getRelPath(execRoot, absPath)
 		if err != nil {
 			return err
 		}
-		fs[normPath] = &fileSysNode{
-			// We cannot directly use meta.Symlink.Target, because it could be
-			// an absolute path. Since the remote worker will map the exec root
-			// to a different directory, we must strip away the local exec root.
-			// See https://github.com/bazelbuild/remote-apis-sdks/pull/229#discussion_r524830458
-			symlink: &symlinkNode{target: relTarget},
-		}
-		if meta.Symlink.IsDangling || !opts.FollowsTarget {
-			return nil
-		}
-		return loadFiles(execRoot, excl, filepath.Clean(filepath.Join(symlinkNormDir, relTarget)), fs, cache, opts)
-	}
-	// Directory
-	files, err := ioutil.ReadDir(absPath)
-	if err != nil {
-		return err
-	}
 
-	if len(files) == 0 {
-		fs[normPath] = &fileSysNode{emptyDirectoryMarker: true}
-		return nil
-	}
-	for _, f := range files {
-		if e := loadFiles(execRoot, excl, filepath.Join(normPath, f.Name()), fs, cache, opts); e != nil {
-			return e
+		switch {
+		// An implication of this is that, if a path is a symlink to a
+		// directory, then the symlink attribute takes precedence.
+		case meta.Symlink != nil && meta.Symlink.IsDangling && !opts.Preserved:
+			// For now, we do not treat a dangling symlink as an error. In the case
+			// where the symlink is not preserved (i.e. needs to be converted to a
+			// file), we simply ignore this path in the finalized tree.
+			continue
+		case meta.Symlink != nil && opts.Preserved:
+			if shouldIgnore(absPath, command.SymlinkInputType, excl) {
+				continue
+			}
+			targetExecRoot, targetSymDir, err := getTargetRelPath(execRoot, normPath, meta.Symlink)
+			if err != nil {
+				return err
+			}
+
+			fs[normPath] = &fileSysNode{
+				// We cannot directly use meta.Symlink.Target, because it could be
+				// an absolute path. Since the remote worker will map the exec root
+				// to a different directory, we must strip away the local exec root.
+				// See https://github.com/bazelbuild/remote-apis-sdks/pull/229#discussion_r524830458
+				symlink: &symlinkNode{target: targetSymDir},
+			}
+
+			if !meta.Symlink.IsDangling && opts.FollowsTarget {
+				// getTargetRelPath validates this target is under execRoot,
+				// and the iteration loop will get the relative path to execRoot,
+				filesToProcess = append(filesToProcess, targetExecRoot)
+			}
+		case meta.IsDirectory:
+			if shouldIgnore(absPath, command.DirectoryInputType, excl) {
+				continue
+			} else if meta.Err != nil {
+				return meta.Err
+			}
+
+			f, err := os.Open(absPath)
```

Comment:
close this?

Reply:
done

Reply:
I think it is better to close earlier.
This may cause 'too many open file' error.

Also why not use ioutil.ReadDir?

Reply:
Fixed to close it earlier.

See my PR description - ioutil.ReadDir isn't efficient as it'll call Stat for files and then unnecessarily sort the results. Besides, ioutil is generally being discouraged from usage: https://golang.org/pkg/io/ioutil/

Reply:
Stat is not used in ReadDir?
https://golang.org/src/io/ioutil/ioutil.go?s=2125:2176#L48

And overhead of sort is ignorable here?

Reply:
If you follow ReadDir's code, `f.Readdir` then calls `readdir(-1, readdirFileInfo)`. The code then calls `lstat`. `readdirname` follows a similar path but doesn't call stat. The sort cost should be minimal but it's still unnecessary wasteful while easy to skip. I'm not sure what's the issue using `ReadDirnames`?

Reply:
Ah, I see. You are right.